### PR TITLE
AirfRANS: Lower LR (5e-4) at 4L/256d with golden config

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The AirfRANS baseline uses lr=7e-4 at 4L/256d. Lower LR was the single biggest breakthrough on TandemFoil (3e-4→2e-4 = -14.7%). We hypothesize that lr=5e-4 at 4L/256d with the full golden config (gc=1.0, WD=1e-2, T_max=5) will provide smoother optimization and potentially deeper convergence than lr=7e-4. The golden config's gradient clipping and weight decay should prevent the instability that higher LRs sometimes cause.

## Instructions
Train on AirfRANS with 4L/256d at lr=5e-4 + golden config:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans \
  --airfrans_task full \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 5 \
  --grad-clip 1.0 \
  --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --epochs 999 \
  --agent kohaku \
  --wandb-name "kohaku/airfrans-4L256d-lr5e4-golden" \
  --wandb-group "airfrans-lr-sweep-golden"
```

CRITICAL FLAGS: --no-use-ema (EMA bug), --epochs 999 (default is 2), SENPAI_MAX_EPOCHS=9999 (default 50 cap), SENPAI_TIMEOUT_MINUTES=180.

Report `val_primary/surface_mse` (lower is better) and note which epoch was best, total epochs, and time per epoch.

## Baseline
Current best: **0.007264** at epoch 40 (4L/256d, gc=1.0, WD=1e-2, T_max=5, lr=7e-4, 50 epochs in 61 min)
PR: #2727 | W&B: ruurxdqs
External target: 0.0043 (1.7x gap)

Reproduce: `cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT kohaku:

## Results

### AirfRANS lr=5e-4 + golden config (4L/256d, gc=1.0, WD=1e-2, T_max=5)

| Metric | Value | vs PR Baseline (0.007264) | vs Radford Best (0.00277) | vs External Target (0.0043) |
|--------|-------|---------------------------|---------------------------|----------------------------|
| **val_primary/surface_mse** | **0.001810** (epoch 184) | **-75.1%** | **-34.7%** | **-57.9%** |
| test_primary/surface_mse | 0.00421 (final epoch 224) | -42.1% | — | -2.1% |

**Val breakdown at best epoch (184):**
- surface_mse_p: 0.007122 (dominant component)
- surface_mse_Ux: 0.000033
- surface_mse_Uy: 0.000039
- surface_mse_nut: 0.000047
- volume_mse: 0.010462

**Test breakdown (final epoch 224, cosine peak — NOT at best val epoch):**
- full_test/surface_mse: 0.00421
- full_test/surface_mse_p: 0.01680
- full_test/volume_mse: 0.01156

### Training details
- **Epochs:** 224 (~0.8 min/epoch, 180-min budget)
- **Best epoch:** 184 (cosine trough, LR=5e-4)
- **Sub-0.003 troughs:** 13 across the run (epochs 118, 121, 122, 128, 131, 144, 146, 156, 182, 184, 185, 191, 195+)
- **Grad norm progression:** 12.0 → 3.2 (very stable)
- **Grad clip events:** 358/360 (99%) early → 170/360 (47%) at best epoch
- **W&B run:** vbtnwitd
- **Peak memory:** Not logged (process completed, GPU freed)

### Command
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw --lr 5e-4 \
  --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --agent kohaku \
  --wandb-name "kohaku/airfrans-4L256d-lr5e4-golden" \
  --wandb-group "airfrans-lr-sweep-golden"
```

### What happened

**lr=5e-4 is a clear winner over lr=7e-4.** The lower learning rate with golden config (gc=1.0, WD=1e-2, T_max=5) produced dramatically better convergence:

1. **Phase transition was earlier and smoother** — the model broke through the plateau around epoch 25-30 (val dropped from 0.04 to 0.018), similar timing to lr=7e-4 but with a smoother trajectory.
2. **Deep basin exploration continued for 200+ epochs** — lr=5e-4 found 13 sub-0.003 troughs vs the previous best of a single 0.00277 trough at gc=0.5. The trough envelope kept descending: 0.00367→0.00330→0.00206→0.00181.
3. **Grad norms stabilized dramatically** — from 12.0 early to 3.2 at best epoch. Clip events dropped from 99% to 47%, meaning the model is in a smooth, well-conditioned basin.
4. **Note on test metric:** Test was evaluated at the final epoch 224 (cosine peak), not at the best val epoch 184. The 0.00421 test metric is likely pessimistic — at a val trough the test would be better.

This mirrors the TandemFoil finding where lower LR was the single biggest lever (-14.7% from 3e-4→2e-4). AirfRANS shows an even more dramatic effect.

### Suggested follow-ups
1. **lr=5e-4 + gc=0.5** — The current radford best uses gc=0.5. Combining with lr=5e-4 could push even deeper since both lower LR and tighter clipping independently improve convergence.
2. **lr=4e-4 or lr=3e-4** — Continue the LR sweep downward; the trend strongly suggests lower LR = deeper basins on AirfRANS.
3. **Checkpoint at cosine trough for test evaluation** — The test metric at final epoch is misleading due to cosine peak. Need trough-aware checkpointing to get true test performance.
4. **Longer training budget** — 224 epochs and still finding new bests at epoch 184. A 360-min budget could push significantly deeper.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2800 because ADVISOR: Closing — your 4L/256d lr=5e-4 result of 0.001810 is solid but does not beat the 0.001479 baseline achieved by 3L/256d. The depth reduction to 3L is the key breakthrough. Good run, kohaku..
